crates/lib/src/api.rs

```diff
@@ -311,3 +319,55 @@ pub async fn retrieve_results(
         .map(ExecutionResults::from)
         .ok_or_else(|| GrpcClientError::ResponseEmpty("Controller Job Execution Results".into()))
 }
+
+/// API Errors encountered when trying to list available quantum processors.
+#[derive(Debug, thiserror::Error)]
+pub enum ListQuantumProcessorNamesError {
+    /// Failed an http call
+    #[error("Failed an http call: {0}")]
+    ApiError(#[from] OpenAPIError<ListQuantumProcessorsError>),
+
+    /// Pagination did not finish before timeout
+    #[error("API pagination did not finish before timeout: {0:?}")]
+    TimeoutError(Duration),
+}
+
+/// Query the QCS API for the names of all available quantum processors.
+/// If `None`, the default `timeout` used is 10 seconds.
+pub async fn list_quantum_processor_names(
+    client: &Qcs,
+    timeout: Option<Duration>,
+) -> Result<Vec<String>, ListQuantumProcessorNamesError> {
+    let timeout = timeout.unwrap_or_else(|| Duration::from_secs(10));
+    let end_time = SystemTime::now() + timeout;
+
+    let mut quantum_processors = vec![];
+    let mut page_token = None;
+
+    loop {
+        if end_time > SystemTime::now() {
+            return Err(ListQuantumProcessorNamesError::TimeoutError(timeout));
+        }
+
+        let result = list_quantum_processors(
```

Comment:
It looks like we only effectively check the timeout between calls to the API, but the API call itself could take an arbitrarily long amount of time since it doesn't have it's own timeout. This would lead to us disrespecting the timeout value.

I think we need to wrap this loop in a construct like [tokio::timer::Timeout](https://docs.rs/tokio/0.1.22/tokio/timer/struct.Timeout.html) with the provided timeout.

Reply:
Adding on to this, I think the best thing to do would be wrap most of the body of the function in an async block and pass the entire thing to `timeout`, i.e.

```rust
let timeout = timeout.unwrap_or_else(|| Duration::from_secs(10));

tokio::time::timeout(timeout, async move {
    let mut quantum_processors = vec![];
    let mut page_token = None;
    loop {
        // .. snip ...
   }

   Ok(quantum_processors)
}).await
```

That way, we don't need to pass a timeout specifically to the API call and don't need to calculate a new timeout for the 2nd+ iterations of the loop.

The only argument against this that I can think of is if we decide we want to return a partial list on timeout, which doesn't seem to be the case here.

Reply:
Ah, I wasn't sure this would actually cancel the underlying loop or just stop waiting for it - prefer to do it that way 👍